### PR TITLE
Get summary attributes for events

### PR DESCRIPTION
**Pull request by josephriosIO**
## What was changed
Gets summary events for for summary view

### CODE REVIEW

File: src/lib/utilities/find-value-if-it-includes-word.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export const findValueIfItIncludesWord = (object, word) => {
+    for (var property in object) {
+      if (object.hasOwnProperty(property) && 
+         property.toString().includes(word)) {
+         return object[property];
+      }
+    }
+  }
```
Comment (GiantRobots) [inline src/lib/utilities/find-value-if-it-includes-word.ts]: No newline on EOF

https://thoughtbot.com/blog/no-newline-at-end-of-file

Talks about why it matters, and also solutions for various editors.

File: src/lib/utilities/history-summary-plucker.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+const getHistorySummary = (event) => {
+  const eventDetails = findValueIfItIncludesWord(event, 'EventAttributes');
+  return isKnownEventType(event.eventType)
+    ? summaryExtractors[event.eventType](eventDetails)
+    : eventDetails;
+};
+
+export { getHistorySummary };
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:93]: I think this file needs prettier ran on it.

File: src/lib/utilities/history-summary-plucker.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+const isKnownEventType = (eventType) => {
+  return eventType in summaryExtractors;
+};
+
+const getHistorySummary = (event) => {
+  const eventDetails = findValueIfItIncludesWord(event, 'EventAttributes');
+  return isKnownEventType(event.eventType)
+    ? summaryExtractors[event.eventType](eventDetails)
+    : eventDetails;
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:90]: You could simplify this line to be 

```typescript
return summaryExtractors[event.eventType](eventDetails) ?? eventDetails
```

The ?? is basically a `if (summaryExtractors[event.eventType](eventDetails)) == null/undefined` then use eventDetails. useful in statements with ?.

File: src/lib/utilities/history-summary-plucker.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    input: d.input,
+    Name: d.activityType.name,
+  }),
+  ActivityTaskStarted: (d) => ({
+    attempt: d.attempt,
+    identity: d.identity,
+    requestId: d.requestId,
+  }),
+  ActivityTaskTimedOut: (d) => ({
+    'Timeout Type': d.failure?.timeoutFailureInfo?.timeoutType,
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:22]: What would the default value for this be? If the `?.` fails it will have undefined as a value so you'll return `{"Timeout Type": undefined}` not sure what the consuming code is like. Will that be okay?

File: src/lib/utilities/history-summary-plucker.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    'Fire Timeout': d.startToFireTimeout?.duration,
+    'Timer ID': d.timerId,
+  }),
+  WorkflowExecutionStarted: (d) => {
+    const summary = {
+      'Close Timeout': d.workflowRunTimeout,
+      identity: d.identity,
+      input: d.input,
+      Parent: undefined,
+      Workflow: d.workflowType ? d.workflowType.name : '',
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:61]: Prefer `d.workflowType?.name ?? ''`

File: src/lib/utilities/history-summary-plucker.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    };
+
+    return summary;
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:64]: Why use a return here instead of the `()=>({})` everywhere else?

File: src/lib/utilities/history-summary-plucker.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import { findValueIfItIncludesWord } from './find-value-if-it-includes-word';
+
+const summaryExtractors = {
+  ActivityTaskCancelRequested: (d) => ({ Id: d.activityId }),
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:4]: I'm not really sure what `d` is in all of these. Can we use a more descriptive name?

Comment (GiantRobots) [reply src/lib/utilities/history-summary-plucker.ts:4]: This code is also super unsafe from a type perspective. All of this would benefit greatly from typing the incoming value 

```typescript
  ActivityTaskCancelRequested: (d:WhateverNetworkValueThisRepresents) => ({ Id: d.activityId }),
```

File: src/lib/utilities/find-value-if-it-includes-word.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+export const findValueIfItIncludesWord = (object, word) => {
+  for (var property in object) {
```
Comment (GiantRobots) [inline src/lib/utilities/find-value-if-it-includes-word.ts:2]: Typically we avoid var these days. 

https://levelup.gitconnected.com/stop-using-var-to-declare-variables-in-javascript-6c0caec16f43

Outlines some basics as to the why.

This is basically just checking if an object has a key with a value in it. You can generally check that like..

```typescript
const myRadObject= {
mojo: "sodope"
};
console.log(myRadObject["mojo"] ?? "Not Dope")
> "sodope"
console.log(myRadObject["Magical"] ?? "Not Dope")
> "Not Dope"
```

File: src/lib/utilities/history-summary-plucker.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  ActivityTaskCompleted: (d) => ({ result: d.result }),
```
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts:5]: For a bunch of these, we can probably have a utility method like `pick`.

```js
const pick = (source, ...keys) => keys.reduce((result, key) => {
  result[key] = source[key];
  return result;
}, {});
```

Or—if you prefer:

```js
const pick = (source, ...keys) => {
  const result = {};
  for (const key of keys) {
    result[key] = source[key];
  };
  return result;
}
```

Then, you could just do something like:

```js
const summaryExtractors = {
  // …
  ActivityTaskCompleted: (d) => pick(d, 'result'),
  ActivityTaskFailed: (d) => pick('details', 'reason'),
  // …
}
```

Basically, we want to avoid having bugs do to typos.

You could add some additional type safety even…

```ts
function pick<T, K extends keyof T>(source: T, ...keys: K[]): Pick<T, K> {
  const result: any = {};
  for (const key of keys) {
    result[key] = source[key];
  }
  return result;
}
```

This will stop us is from accidentally typing a key that does not exist on the source object.

There is probably even fancier stuff we could do with currying but the gymnastics we'd need to do with types probably isn't worth it.

File: src/lib/utilities/find-value-if-it-includes-word.ts (hunk 8 above)
Comment (GiantRobots) [reply src/lib/utilities/find-value-if-it-includes-word.ts:2]: This can also probably be avoided in your code if you have types in your code.

File: src/lib/utilities/history-summary-plucker.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  ChildWorkflowExecutionStarted: (d) => ({
+    Workflow: d,
+  }),
+  WorkflowTaskCompleted: (d) => ({ identity: d.identity }),
+  WorkflowTaskScheduled: (d) => ({
+    Taskqueue: d.taskQueue.name,
+    Timeout: d.startToCloseTimeout,
+  }),
+  WorkflowTaskStarted: (d) => ({ requestId: d.requestId }),
+  WorkflowTaskTimedOut: (d) => ({ 'Timeout Type': d.timeoutType }),
```
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts:37]: I don't think we need to worry about the inconsistent capitalization from the current UI. We can just pick properties as they show up in the JSON.

File: src/lib/utilities/history-summary-plucker.ts (hunk 6 above)
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts:64]: We can probably just return the object like the rest.

File: src/lib/utilities/history-summary-plucker.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    return summary;
+  },
+  WorkflowExecutionCompleted: (d) => ({
+    result: d.result,
+    newExecutionRunId: d.newExecutionRunId || undefined,
+  }),
+  WorkflowExecutionFailed: (d) => ({
+    message: d.failure?.message,
+    newExecutionRunId: d.newExecutionRunId || undefined,
```
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts:72]: Wouldn't this be `undefined` anyway?

File: src/lib/utilities/history-summary-plucker.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  }),
+  WorkflowExecutionTimedOut: (d) => ({
+    retryState: d.retryState,
+    newExecutionRunId: d.newExecutionRunId || undefined,
+  }),
+  WorkflowTaskFailed: (d) => ({ message: d.failure?.message }),
+  ChildWorkflowExecutionFailed: (d) => ({ message: d.failure?.message }),
+};
+
+const isKnownEventType = (eventType) => {
```
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts]: TypeScript should be able to handle this for us.

File: src/lib/utilities/history-summary-plucker.ts (hunk 2 above)
Comment (stevekinney) [inline src/lib/utilities/history-summary-plucker.ts:87]: You can probably just do `export const` on Line 86.

File: src/lib/utilities/find-value-if-it-includes-word.ts (hunk 8 above)
Comment (GiantRobots) [reply src/lib/utilities/find-value-if-it-includes-word.ts:2]: Also generally avoid `for( .. in ... )`. Prefer `for .. of` because it doesn't enumerate over object fields

https://stackoverflow.com/questions/500504/why-is-using-for-in-for-array-iteration-a-bad-idea

For of docs

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

Also also looking up objects this way is `O(N)` ie. it will iterate over every item in an object to find that instead of `O(1)` which will access the key directly. 

To put it another way. If you have an object with 10,000 keys you can do 10,000 checks or you can do 1 check depending on how you access the object.

File: src/lib/utilities/find-value-if-it-includes-word.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+export const findValueIfItIncludesWord = (object, word) => {
```
Comment (stevekinney) [inline src/lib/utilities/find-value-if-it-includes-word.ts:1]: We already have a more TypeScript-friend (if also more verbose) version of this in `src/lib/utilities/get-component-for-event-type.ts`.

File: src/lib/utilities/history-summary-plucker.ts (hunk 5 above)
Comment (stevekinney) [reply src/lib/utilities/history-summary-plucker.ts:61]: @GiantRobots taught me about this today. I literally didn't know it existed.

File: src/lib/utilities/history-summary-plucker.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  WorkflowExecutionTimedOut: (d) => ({
+    retryState: d.retryState,
+    newExecutionRunId: d.newExecutionRunId || undefined,
+  }),
+  WorkflowTaskFailed: (d) => ({ message: d.failure?.message }),
+  ChildWorkflowExecutionFailed: (d) => ({ message: d.failure?.message }),
+};
+
+const getHistorySummary = (event) => {
+  const eventDetails = findValueIfItIncludesWord(event, 'EventAttributes');
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.ts:83]: eventDetails can be undefined here which may cause errors when using it with other functions

**Review by GiantRobots — CHANGES_REQUESTED** [review]
history-summary-plucker is really difficult to understand without types and we should probably add a bunch

File: src/lib/utilities/history-summary-plucker.ts (hunk 9 above)
Comment (GiantRobots) [reply src/lib/utilities/history-summary-plucker.ts:5]: You can also do picks with destructuring like...

```typescript
  ActivityTaskFailed: ({details, reason}) => ({details, reason})
```

Although pick may be nicer.

File: src/lib/utilities/history-summary-plucker.test.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    };
+
+    expect(getHistorySummary(event)).toStrictEqual({
+      'Close Timeout': '100s',
+      identity: 'some identity',
+      input: 'true',
+      Parent: undefined,
+      Workflow: 'Rainbow Workflow',
+    });
+  });
```
Comment (GiantRobots) [inline src/lib/utilities/history-summary-plucker.test.ts:26]: When you add types to everything you can check if there are type errors with ts expect error comments

https://devblogs.microsoft.com/typescript/announcing-typescript-3-9-beta/#ts-expect-error-comments

**Review by GiantRobots — CHANGES_REQUESTED** [review]

**Review by stevekinney — APPROVED** [review]
